src/bisimulation.rs

```diff
@@ -0,0 +1,372 @@
+) {
+    let mut all_new_fine_blocks: Vec<Rc<RefCell<FineBlock>>> = vec![];
+    let mut all_removed_fine_blocks: Vec<Rc<RefCell<FineBlock>>> = vec![];
+    let mut new_compound_coarse_blocks: Vec<Rc<CoarseBlock>> = vec![];
+
+    for (block, counter_image_group) in counterimage_grouped.iter_mut() {
+        let borrowed_coarse_block = Rc::clone(
+            &(*counter_image_group.block)
+                .borrow()
+                .coarse_block_that_supersets_self,
+        );
+
+        let proper_subblock = {
+            let fine_block = FineBlock {
+                values: counter_image_group.subblock.clone(),
+                coarse_block_that_supersets_self: Rc::clone(&borrowed_coarse_block),
+            };
+
+            Rc::new(RefCell::new(fine_block))
+        };
+        let prior_count = borrowed_coarse_block.fine_block_count();
+        borrowed_coarse_block.add_fine_block(Rc::clone(&proper_subblock));
+
+        if prior_count == 1 {
+            new_compound_coarse_blocks.push(Rc::clone(&borrowed_coarse_block));
+        }
+
+        for node_index in counter_image_group.subblock.iter() {
+            node_to_block_vec[node_index.index()] = Rc::clone(&proper_subblock);
+        }
+
+        // subtract subblock from block
+        (*counter_image_group.block).borrow_mut().values = block
+            .iter()
+            .filter(|x| !(*proper_subblock).borrow().values.contains(x))
+            .copied()
+            .collect();
+
+        if (*counter_image_group.block).borrow().values.is_empty() {
+            borrowed_coarse_block.remove_fine_block(&counter_image_group.block);
+            all_removed_fine_blocks.push(Rc::clone(&counter_image_group.block));
+        }
+        all_new_fine_blocks.push(Rc::clone(&proper_subblock));
+    }
+    (
+        (all_new_fine_blocks, all_removed_fine_blocks),
+        new_compound_coarse_blocks,
+    )
+}
+
+fn maximum_bisimulation(graph: &StablePyGraph<Directed>) -> Vec<Block> {
+    let (fine_block_tuple, initial_coarse_partition, mut node_to_block_vec) = initialization(graph);
+
+    let mut queue: CoarsePartition = initial_coarse_partition;
+    let mut all_fine_blocks = vec![fine_block_tuple.0, fine_block_tuple.1];
+
+    while !queue.is_empty() {
+        let (smaller_component, simple_splitter_block) = {
+            let splitter_block = queue.pop().unwrap();
+            let mut fine_blocks_in_splitter_block = splitter_block
+                .fine_blocks_that_are_subsets_of_self
+                .borrow()
+                .clone();
+
+            let smaller_component_index = fine_blocks_in_splitter_block
+                .iter()
+                .enumerate()
+                .min_by(|(_, x), (_, y)| {
+                    (***x)
+                        .borrow()
+                        .values
+                        .len()
+                        .cmp(&(***y).borrow().values.len())
+                })
+                .map(|(index, _)| index)
+                .unwrap();
```

Comment:
Instead of `.unwrap()`ing, make the function return a result and use `?` to return the error to the caller. When Rust panics the experience for Python users is not good